File: javascript/binary-search-tree/binary-search-tree.js
'use strict';

function Tree(value) {
    this.data  = value;
    this.right = null;
    this.left  = null;
}

Tree.prototype.insert = function (value) {
    var side = value <= this.data ? 'left' : 'right';
    if (this[side] === null) {
        this[side] = new Tree(value);
    } else {
        this[side].insert(value);
    }
};

Tree.prototype.each = function (cb) {
    if (this.left) {
        this.left.each(cb); // visit the left
    }
    cb.call(null, this.data); // visit itself
    if (this.right) {
        return this.right.each(cb); // visit the right
    }
};

module.exports = Tree;

File: elixir/binary-search-tree/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %{
      data: data,
      left: nil,
      right: nil
    }
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(%{left: nil, data: node_data} = tree, data)
      when node_data >= data do
    %{
      tree
      | left: new(data)
    }
  end

  def insert(%{right: nil, data: node_data} = tree, data)
      when node_data < data do
    %{
      tree
      | right: new(data)
    }
  end

  def insert(%{left: left, data: node_data} = tree, data)
      when node_data >= data do
    %{tree | left: insert(left, data)}
  end

  def insert(%{right: right, data: node_data} = tree, data)
      when node_data < data do
    %{tree | right: insert(right, data)}
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(nil), do: []

  def in_order(tree) do
    in_order(tree.left) ++ [tree.data | in_order(tree.right)]
  end
end
